File: 1_50/23_merge_k_sorted_list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *res = malloc(sizeof(struct ListNode));
    struct ListNode *p = res;
    while ((l1) && (l2)) {
        if (l1->val < l2->val) {
            p->next = l1;
            l1 = l1->next;
        } else {
            p->next = l2;
            l2 = l2->next;
        }
        p = p->next;
    }
    // add the remaining values
    p->next = l1 ? l1 : l2;
    return res->next;
}

struct ListNode* mergeKLists(struct ListNode** lists, int listsSize) {
    if (lists == NULL || listsSize == 0) {
        return NULL;
    } else if (listsSize == 1) {
        return lists[0];
    }
    
    int last = listsSize - 1;
    
    while (last > 0) {
        // merge first and last
        int i = 0, j = last;
        
        // (lists[i], lists[j]) is a pair
        while (i < j) {
            lists[i] = mergeTwoLists(lists[i], lists[j]);
            i++;
            j--;
        }
        
        // update index of last
        last = j;
    }
    // return the one remaining
    return lists[0];
}

File: 1_50/23_merge_k_sorted_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

import java.util.PriorityQueue;

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        } else if (lists.length == 1) {
            return lists[0];
        }
        
        ListNode res = new ListNode(0);
        ListNode p = res;
        
        PriorityQueue<ListNode> queue = new PriorityQueue<>(new Comparator<ListNode>(){
            public int compare(ListNode l1, ListNode l2) {
                return l1.val - l2.val;
            }
        });
        
        for (ListNode list : lists) {
            if (list != null) {
                queue.add(list);
            }
        }
        
        while (!queue.isEmpty()) {
            ListNode head = queue.poll();
            p.next = head;
            p = p.next;
            if (head.next != null) {
                queue.add(head.next);
            }
        }
        
        return res.next;
    }
}
